File: ideviewer/gitleaks.py
"""
Gitleaks installer and manager.

Handles installation and version detection of gitleaks for pre-commit secret scanning.
"""

import json
import logging
import os
import platform
import shutil
import stat
import subprocess
import sys
import tarfile
import tempfile
import zipfile
from pathlib import Path
from typing import Optional
from urllib.request import Request, urlopen
from urllib.error import URLError, HTTPError

logger = logging.getLogger(__name__)

GITHUB_RELEASES_URL = "https://api.github.com/repos/gitleaks/gitleaks/releases/latest"
IDEVIEWER_BIN_DIR = Path.home() / ".ideviewer" / "bin"


def _get_gitleaks_binary_name() -> str:
    """Get the gitleaks binary name for this platform."""
    if sys.platform == "win32":
        return "gitleaks.exe"
    return "gitleaks"


def _get_ideviewer_gitleaks_path() -> Path:
    """Get the path to the ideviewer-managed gitleaks binary."""
    return IDEVIEWER_BIN_DIR / _get_gitleaks_binary_name()


def is_gitleaks_installed() -> bool:
    """Check if gitleaks is available (either in ideviewer bin or system PATH)."""
    # Check ideviewer-managed binary first
    ideviewer_path = _get_ideviewer_gitleaks_path()
    if ideviewer_path.exists() and os.access(str(ideviewer_path), os.X_OK):
        return True

    # Check system PATH
    return shutil.which("gitleaks") is not None


def get_gitleaks_version() -> Optional[str]:
    """Get the installed gitleaks version string, or None if not installed."""
    # Check ideviewer-managed binary first
    ideviewer_path = _get_ideviewer_gitleaks_path()
    gitleaks_cmd = None

    if ideviewer_path.exists() and os.access(str(ideviewer_path), os.X_OK):
        gitleaks_cmd = str(ideviewer_path)
    elif shutil.which("gitleaks"):
        gitleaks_cmd = "gitleaks"

    if not gitleaks_cmd:
        return None

    try:
        result = subprocess.run(
            [gitleaks_cmd, "version"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        if result.returncode == 0:
            return result.stdout.strip()
        return None
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        return None


def _get_platform_asset_pattern() -> str:
    """Get the expected asset filename pattern for this platform."""
    system = platform.system().lower()
    machine = platform.machine().lower()

    # Normalize architecture names
    if machine in ("x86_64", "amd64"):
        arch = "x64"
    elif machine in ("aarch64", "arm64"):
        arch = "arm64"
    elif machine in ("i386", "i686", "x86"):
        arch = "x32"
    else:
        arch = machine

    if system == "darwin":
        return f"gitleaks_{{}}_darwin_{arch}.tar.gz"
    elif system == "linux":
        return f"gitleaks_{{}}_linux_{arch}.tar.gz"
    elif system == "windows":
        return f"gitleaks_{{}}_windows_{arch}.zip"
    else:
        return f"gitleaks_{{}}_linux_{arch}.tar.gz"


def _download_from_github() -> bool:
    """Download the latest gitleaks release from GitHub."""
    logger.info("Downloading gitleaks from GitHub releases...")

    try:
        req = Request(
            GITHUB_RELEASES_URL,
            headers={"User-Agent": "IDEViewer/0.1.0"},
        )
        with urlopen(req, timeout=30) as response:
            release_data = json.loads(response.read().decode("utf-8"))
    except (URLError, HTTPError, json.JSONDecodeError) as e:
        logger.error(f"Failed to fetch GitHub release info: {e}")
        return False

    version = release_data.get("tag_name", "").lstrip("v")
    assets = release_data.get("assets", [])

    if not version or not assets:
        logger.error("No release version or assets found")
        return False

    # Find the right asset for this platform
    system = platform.system().lower()
    machine = platform.machine().lower()

    # Normalize architecture
    if machine in ("x86_64", "amd64"):
        arch = "x64"
    elif machine in ("aarch64", "arm64"):
        arch = "arm64"
    elif machine in ("i386", "i686", "x86"):
        arch = "x32"
    else:
        arch = machine

    target_name = None
    download_url = None

    for asset in assets:
        name = asset.get("name", "").lower()
        # Match platform and architecture
        if system == "darwin" and "darwin" in name and arch in name:
            target_name = asset["name"]
            download_url = asset.get("browser_download_url")
            break
        elif system == "linux" and "linux" in name and arch in name and not name.endswith(".rpm") and not name.endswith(".deb"):
            target_name = asset["name"]
            download_url = asset.get("browser_download_url")
            break
        elif system == "windows" and "windows" in name and arch in name:
            target_name = asset["name"]
            download_url = asset.get("browser_download_url")
            break

    if not download_url:
        logger.error(f"No suitable gitleaks binary found for {system}/{arch}")
        return False

    logger.info(f"Downloading {target_name}...")

    try:
        req = Request(
            download_url,
            headers={"User-Agent": "IDEViewer/0.1.0"},
        )
        with urlopen(req, timeout=120) as response:
            archive_data = response.read()
    except (URLError, HTTPError) as e:
        logger.error(f"Failed to download gitleaks: {e}")
        return False

    # Create bin directory
    IDEVIEWER_BIN_DIR.mkdir(parents=True, exist_ok=True)

    binary_name = _get_gitleaks_binary_name()
    dest_path = _get_ideviewer_gitleaks_path()

    try:
        with tempfile.TemporaryDirectory() as tmpdir:
            tmpdir_path = Path(tmpdir)
            archive_path = tmpdir_path / target_name

            archive_path.write_bytes(archive_data)

            if target_name.endswith(".tar.gz") or target_name.endswith(".tgz"):
                with tarfile.open(archive_path, "r:gz") as tar:
                    tar.extractall(path=tmpdir, filter="data")
            elif target_name.endswith(".zip"):
                with zipfile.ZipFile(archive_path, "r") as zf:
                    zf.extractall(path=tmpdir)
            else:
                logger.error(f"Unknown archive format: {target_name}")
                return False

            # Find the gitleaks binary in extracted files
            extracted_binary = None
            for root, dirs, files in os.walk(tmpdir):
                for f in files:
                    if f == binary_name or f == "gitleaks":
                        extracted_binary = Path(root) / f
                        break
                if extracted_binary:
                    break

            if not extracted_binary or not extracted_binary.exists():
                logger.error("Could not find gitleaks binary in archive")
                return False

            # Copy to destination
            shutil.copy2(str(extracted_binary), str(dest_path))

            # Make executable on Unix
            if sys.platform != "win32":
                dest_path.chmod(dest_path.stat().st_mode | stat.S_IEXEC | stat.S_IXGRP | stat.S_IXOTH)

    except Exception as e:
        logger.error(f"Failed to extract gitleaks: {e}")
        return False

    logger.info(f"gitleaks installed to {dest_path}")
    return True


def _install_via_brew() -> bool:
    """Install gitleaks via Homebrew on macOS."""
    if not shutil.which("brew"):
        return False

    logger.info("Installing gitleaks via Homebrew...")
    try:
        result = subprocess.run(
            ["brew", "install", "gitleaks"],
            capture_output=True,
            text=True,
            timeout=300,
        )
        if result.returncode == 0:
            logger.info("gitleaks installed via Homebrew")
            return True
        else:
            logger.warning(f"Homebrew install failed: {result.stderr.strip()}")
            return False
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError) as e:
        logger.warning(f"Homebrew install failed: {e}")
        return False


def install_gitleaks() -> bool:
    """
    Install gitleaks. Returns True if installed successfully.

    Installation methods by platform:
    - macOS: brew install gitleaks (if brew available), otherwise GitHub release
    - Linux: GitHub release binary
    - Windows: GitHub release binary
    """
    if is_gitleaks_installed():
        version = get_gitleaks_version()
        logger.info(f"gitleaks already installed (version: {version})")
        return True

    system = platform.system().lower()

    # macOS: try Homebrew first
    if system == "darwin":
        if _install_via_brew():
            return True
        logger.info("Homebrew not available or failed, falling back to GitHub release")

    # All platforms: download from GitHub
    return _download_from_github()

File: pkg/gitleaks/gitleaks.go
// Package gitleaks manages installation and version detection of the gitleaks binary.
package gitleaks

import (
	"archive/tar"
	"archive/zip"
	"bytes"
	"compress/gzip"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/securient/ideviewer-oss/internal/platform"
)

const githubReleasesURL = "https://api.github.com/repos/gitleaks/gitleaks/releases/latest"

// IsInstalled returns true if gitleaks is available (ideviewer-managed or system PATH).
func IsInstalled() bool {
	binPath := platform.GitleaksBinPath()
	if info, err := os.Stat(binPath); err == nil && !info.IsDir() {
		return true
	}
	_, err := exec.LookPath("gitleaks")
	return err == nil
}

// GetVersion returns the installed gitleaks version string, or an error.
func GetVersion() (string, error) {
	gitleaksCmd := ""

	binPath := platform.GitleaksBinPath()
	if info, err := os.Stat(binPath); err == nil && !info.IsDir() {
		gitleaksCmd = binPath
	} else if p, err := exec.LookPath("gitleaks"); err == nil {
		gitleaksCmd = p
	}

	if gitleaksCmd == "" {
		return "", fmt.Errorf("gitleaks not found")
	}

	out, err := exec.Command(gitleaksCmd, "version").Output()
	if err != nil {
		return "", fmt.Errorf("failed to get gitleaks version: %w", err)
	}
	return strings.TrimSpace(string(out)), nil
}

// Install downloads and installs gitleaks. On macOS it tries brew first.
func Install() error {
	if IsInstalled() {
		return nil // already installed
	}

	// macOS: try Homebrew first.
	if runtime.GOOS == "darwin" {
		if err := installViaBrew(); err == nil {
			return nil
		}
	}

	return downloadFromGitHub()
}

// installViaBrew attempts to install gitleaks via Homebrew.
func installViaBrew() error {
	if _, err := exec.LookPath("brew"); err != nil {
		return fmt.Errorf("brew not found")
	}
	cmd := exec.Command("brew", "install", "gitleaks")
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

// normalizedArch returns the gitleaks asset architecture string.
func normalizedArch() string {
	switch runtime.GOARCH {
	case "amd64":
		return "x64"
	case "arm64":
		return "arm64"
	case "386":
		return "x32"
	default:
		return runtime.GOARCH
	}
}

// githubRelease is the subset of the GitHub release API we need.
type githubRelease struct {
	TagName string        `json:"tag_name"`
	Assets  []githubAsset `json:"assets"`
}

type githubAsset struct {
	Name               string `json:"name"`
	BrowserDownloadURL string `json:"browser_download_url"`
}

// downloadFromGitHub downloads the latest gitleaks release for this platform.
func downloadFromGitHub() error {
	client := &http.Client{Timeout: 30 * time.Second}
	req, err := http.NewRequest("GET", githubReleasesURL, nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", "IDEViewer")

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to fetch GitHub release info: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("GitHub API returned %d", resp.StatusCode)
	}

	var release githubRelease
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		return fmt.Errorf("failed to parse GitHub release: %w", err)
	}

	if release.TagName == "" || len(release.Assets) == 0 {
		return fmt.Errorf("no release version or assets found")
	}

	// Find matching asset.
	system := runtime.GOOS
	arch := normalizedArch()

	var targetAsset *githubAsset
	for i, asset := range release.Assets {
		name := strings.ToLower(asset.Name)
		switch system {
		case "darwin":
			if strings.Contains(name, "darwin") && strings.Contains(name, arch) {
				targetAsset = &release.Assets[i]
			}
		case "linux":
			if strings.Contains(name, "linux") && strings.Contains(name, arch) &&
				!strings.HasSuffix(name, ".rpm") && !strings.HasSuffix(name, ".deb") {
				targetAsset = &release.Assets[i]
			}
		case "windows":
			if strings.Contains(name, "windows") && strings.Contains(name, arch) {
				targetAsset = &release.Assets[i]
			}
		}
		if targetAsset != nil {
			break
		}
	}

	if targetAsset == nil {
		return fmt.Errorf("no suitable gitleaks binary found for %s/%s", system, arch)
	}

	// Download the asset.
	dlClient := &http.Client{Timeout: 120 * time.Second}
	dlReq, err := http.NewRequest("GET", targetAsset.BrowserDownloadURL, nil)
	if err != nil {
		return err
	}
	dlReq.Header.Set("User-Agent", "IDEViewer")

	dlResp, err := dlClient.Do(dlReq)
	if err != nil {
		return fmt.Errorf("failed to download gitleaks: %w", err)
	}
	defer dlResp.Body.Close()

	archiveData, err := io.ReadAll(dlResp.Body)
	if err != nil {
		return fmt.Errorf("failed to read download: %w", err)
	}

	// Create bin directory.
	binDir := platform.GitleaksBinDir()
	if err := os.MkdirAll(binDir, 0o755); err != nil {
		return fmt.Errorf("failed to create bin directory: %w", err)
	}

	binaryName := "gitleaks"
	if runtime.GOOS == "windows" {
		binaryName = "gitleaks.exe"
	}

	destPath := platform.GitleaksBinPath()

	// Extract based on archive type.
	assetName := strings.ToLower(targetAsset.Name)
	if strings.HasSuffix(assetName, ".tar.gz") || strings.HasSuffix(assetName, ".tgz") {
		if err := extractTarGz(archiveData, binaryName, destPath); err != nil {
			return err
		}
	} else if strings.HasSuffix(assetName, ".zip") {
		if err := extractZip(archiveData, binaryName, destPath); err != nil {
			return err
		}
	} else {
		return fmt.Errorf("unknown archive format: %s", targetAsset.Name)
	}

	// Make executable on Unix.
	if runtime.GOOS != "windows" {
		if err := os.Chmod(destPath, 0o755); err != nil {
			return fmt.Errorf("failed to set executable permission: %w", err)
		}
	}

	return nil
}

// extractTarGz extracts the named binary from a tar.gz archive.
func extractTarGz(data []byte, binaryName, destPath string) error {
	gz, err := gzip.NewReader(bytes.NewReader(data))
	if err != nil {
		return fmt.Errorf("failed to open gzip: %w", err)
	}
	defer gz.Close()

	tr := tar.NewReader(gz)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return fmt.Errorf("tar read error: %w", err)
		}
		base := filepath.Base(hdr.Name)
		if base == binaryName || base == "gitleaks" {
			f, err := os.Create(destPath)
			if err != nil {
				return fmt.Errorf("failed to create binary file: %w", err)
			}
			if _, err := io.Copy(f, tr); err != nil {
				f.Close()
				return fmt.Errorf("failed to write binary: %w", err)
			}
			f.Close()
			return nil
		}
	}
	return fmt.Errorf("gitleaks binary not found in archive")
}

// extractZip extracts the named binary from a zip archive.
func extractZip(data []byte, binaryName, destPath string) error {
	r, err := zip.NewReader(bytes.NewReader(data), int64(len(data)))
	if err != nil {
		return fmt.Errorf("failed to open zip: %w", err)
	}
	for _, f := range r.File {
		base := filepath.Base(f.Name)
		if base == binaryName || base == "gitleaks" {
			rc, err := f.Open()
			if err != nil {
				return fmt.Errorf("failed to open zip entry: %w", err)
			}
			defer rc.Close()
			out, err := os.Create(destPath)
			if err != nil {
				return fmt.Errorf("failed to create binary file: %w", err)
			}
			if _, err := io.Copy(out, rc); err != nil {
				out.Close()
				return fmt.Errorf("failed to write binary: %w", err)
			}
			out.Close()
			return nil
		}
	}
	return fmt.Errorf("gitleaks binary not found in zip archive")
}
